### DISCUSSION TITLE
[Proposal]:  "Guard" method parameters

### DISCUSSION THREAD

**Discussion by mhomde** [dotnet/csharplang#5625] [Language Ideas]
# Guard method parameters

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Be able to mark method parameters as non-null, otherwise, a method will not run,  and/or return the default of its return type.

## Motivation
[motivation]: #motivation

Remove cluttering null-checking "guard statements" from the start of methods

## Detailed design
[design]: #detailed-design

So one of the most common clutter I still encounter in day-to-day coding, despite all null-related features that have been added in the last few years, is starting methods with one or several guard statements that check if something is null, and if they fail return immediately. 

```
public ReturnObject SomeMethod(object a, object b)
{
   if (a == null)
    return null;
    ...
}
```

Sure there are non-nullable reference types, but they are enabled on per-project basis and is more for compiler checks and recommendation, and not runtime (afaik). You can't override per parameter as nullable. With this feature, you could more easily clean up legacy code. 

The idea would be to just mark these parameters with !  in the method header like  so:

```
public ReturnObject SomeMethod(object! a, object b)
{
  ...
}
```

If it fails it returns the default of the return type, so in this case null. 

### Guard overrides
You could also have different "overrides" (maybe, might introduce too much complexity)

```
public ReturnObject SomeMethod(object! a, object! b)
{
  ....
}

public ReturnObject SomeMethod(object a, object! b)
{
  ....
}
```

###  Fail handling
An extension is that you'll probably want to handle the case where the guard fails, for logging, returning a different value, etc. This could be solved by falling back to a method override without guard parameters

```
public ReturnObject SomeMethod(object a, object b)
{
   LogSomeError();
   return new ErrorReturnObject();
}
```

## Drawbacks
[drawbacks]: #drawbacks

Is the reduced clutter worth the language feature? Does it provide enough benefits with nullable-reference types being available? Are there edge-cases of return types that wouldn't work? Is there other features the ! syntax would need to be reserved for?

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

It would, perhaps, be nice to mark the "fail" method somehow to show its intent, and perhaps enforce that there are methods with guard statements.  I can't come up with an elegant syntax, however. Something like

```
public ReturnObject SomeMethod(object a, object b) : fail
{
   LogSomeError();
   return new ErrorReturnObject();
}
```

Dunno, is it worth introducing another keyword?, and perhaps it's unecessary, but would make reading the code a bit more clear

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/issues/2145

IMO handling the null case by returning a default value or the like feels very off to me, something I would want to be explicit in the code.

**Reply by mhomde** [reply]
I agree it requires some thought, the "guardless" method could handle return values, but it's not very explicit. I guess the compiler could enforce a "fail" method (if one chose to implement that keyword).

If you have overrides on different guard parameters you can't really put a default fail return type on the individual methods with guard parameters.

That being said, I think returning default(T) could work for simplicity's sake

**Comment by foxesknow** [comment]
Returning the default value for a type feels wrong, and how would it work for methods that return `void` ?

It'd be much better to throw an exception.

**Reply by mhomde** [reply]
void would just not run the method. You might want to throw an exception (this could be configurable I guess), but not necessarily. That's why you have the "fail" fallback override that decides what to do.  

I mean this can become a pretty philosophically discussion about error management and try-catch, but throwing exceptions willy nilly in many projects are often not feasible :)  And I would  say its very common in practice to just return from a method with null if the guard clauses doesn't pass.

**Reply by foxesknow** [reply]
> And I would say its very common in practice to just return from a method with null if the guard clauses doesn't pass

I really don't think that is true.

**Reply by theunrepentantgeek** [reply]
> throwing exceptions willy nilly in many projects are often not feasible

It's not only feasible, it is the normal and idiomatic way to do things in the .Net ecosystem.

In my experience, silently concealing problems by returning default values is a significant and fertile source of data corruption bugs.

**Reply by yaakov-h** [reply]
There's a reason that Apple turned years of Objective-C's `nil` behaviour on its head when they built Swift and went for strict type checking and guard statements instead.

Propagating failures silently is a horrible experience for everyone involved.

**Reply by mhomde** [reply]
It's not all about errors or exceptions, you can have functions that by design (which might be legacy) only runs if something is not null. For example, if you do traversal through an object hierarchy, then you don't continue if a Property is null.

Sure, you could say that it's up to each callee to not pass null, and with things like optionals and nullable reference types you can go for that. But it's not always feasible in real-world projects
